llm/bedrock/src/wasi_client.rs

```diff
@@ -0,0 +1,164 @@
+}
+
+unsafe impl Send for WasiClient {}
+unsafe impl Sync for WasiClient {}
+
+#[derive(Debug)]
+struct SharedWasiConnector {
+    inner: Arc<WasiConnector>,
+}
+
+impl SharedWasiConnector {
+    fn new(client: reqwest::Client) -> Self {
+        Self {
+            inner: Arc::new(WasiConnector(client)),
+        }
+    }
+}
+
+#[derive(Debug)]
+struct WasiConnector(reqwest::Client);
+
+unsafe impl Send for WasiConnector {}
+unsafe impl Sync for WasiConnector {}
+
+impl WasiConnector {
+    async fn handle(
+        &self,
+        request: config::http::HttpRequest,
+    ) -> Result<reqwest::Response, ConnectorError> {
+        let method = Method::from_bytes(request.method().as_bytes()).expect("Valid http method");
+        let url = request.uri().to_owned();
+        let parts = request.into_parts();
+
+        let mut header_map = HeaderMap::new();
+
+        for header in parts.headers.iter() {
+            header_map.append(
+                HeaderName::from_str(header.0).expect("Valid http header name"),
+                HeaderValue::from_str(header.1).expect("Valid http header value"),
+            );
+        }
+
+        self.0
+            .request(method, url)
+            .headers(header_map)
+            .body(reqwest::Body::new::<BodyReader>(parts.body.into()))
+            .send()
+            .await
+            .map_err(|e| ConnectorError::other(e.into(), None))
+    }
+}
+
+impl HttpConnector for SharedWasiConnector {
+    fn call(&self, request: config::http::HttpRequest) -> HttpConnectorFuture {
+        let inner_clone = Arc::clone(&self.inner);
+
+        let future = async move {
+            let response = inner_clone.handle(request).await?;
+            log::trace!("WasiConnector: response received {response:?}");
+
+            let status_code: StatusCode = response.status().into();
+            let headers_map = response.headers().clone();
+            let extensions = response.extensions().clone();
+
+            let body = response
+                .bytes()
+                .await
+                .map(|body| {
+                    if body.is_empty() {
+                        SdkBody::empty()
+                    } else {
+                        SdkBody::from(body)
+                    }
+                })
+                .map_err(|e| ConnectorError::other(e.into(), None))?;
+
+            let mut headers = Headers::new();
+            for header in headers_map {
+                if let Some(key) = header.0 {
+                    if let Ok(value) = header.1.to_str() {
+                        headers.insert(key.to_string(), value.to_string());
+                    }
+                }
+            }
+
+            let mut sdk_response = Response::new(status_code, body);
+            *sdk_response.headers_mut() = headers;
+            sdk_response.add_extension(extensions);
+
+            Ok(sdk_response)
+        };
+
+        HttpConnectorFuture::new(UnsafeFuture::new(future))
+    }
+}
+
+struct BodyReader {
```

Comment:
also tried you `BodyReader` trick, sadly this breaks every api that relies on data integrity check to verify that the data received is the same data that was originally sent. Underlying error you get is `XAmzContentSHA256Mismatch`

Reply:
Seems I need to keep my own custom implementations, really hoped that maybe you cracked it :(

Reply:
How are you doing the verification? I don't get any errors when I run it myself. Bedrock responds correctly without the error you described. Maybe there is something different you are doing?

Reply:
I am testing it with s3, so with this https://github.com/awslabs/aws-sdk-rust/tree/main/sdk/s3. Using s3 it really matters that you do not make a copy of the body as you are working with potentially large files. Why I need s3 is to use aws transcribe you first need to upload your audio to s3

Reply:
Alright I understand. You can consider recomputing the sha256 from the byte reference. You shouldn't need an owned slice for that (I guess?).

I am curious about your own workaround. Mine works for this solution anyways so I think it's good

Reply:
This comes to mind

[https://docs.rs/sha2/latest/sha2/trait.Digest.html#tymethod.digest](https://docs.rs/sha2/latest/sha2/trait.Digest.html#tymethod.digest)

Reply:
This is the HttpClient impl I used with AWS Sdk https://gist.github.com/kmatasfp/ef9cdc09510eaf76fe90d05285911cd6

Reply:
Thanks, but of you look at the gist, you will see I already have a working solution to use parts of aws with just reqwest http client. But I wanted to still use sdk as it means less code we need to maintain ourselves.

> And this is my completely custom aws client impl with tests https://gist.github.com/kmatasfp/c21e224f9f201eb99c47792daadbf2b4

Reply:
@kmatasfp using the `sha2` crate, you can compute the hash again from the slice in SdkBody and check if it's different from the one being supplied in the header map provided by the aws client.

If it is not then you should just use the new one you computed. If it is then I am out of ideas 😄

Reply:
Verified, but as I suspected this is not the issue, but I verified, manually calculated hash matches what is in the headers already. Why I get `XAmzContentSHA256Mismatch` is because the hash does not match on the server side, so basically s3 is receiving different body than I sent. And I think there is a bug in the `reqwest` client itself.  Namely issue arises when you send binary data, that is why you did not encounter this issue, you are only send text, in my case I am sending audio blob

Reply:
Your implementation is in fact the same as what is implemented in std by slice.

That's why I called bytes.read directly.

![image](https://github.com/user-attachments/assets/4a0f41d9-fe6f-4c0b-b96c-7b98fa39ffa2)

Reply:
Yeah this is what I replicated, but notice me remembering the current read `position`. And everything works after that. In your case I think you lose it when we leave the `read` function. `bytes` is local variable, no?  We are getting a new borrow every time we come to `read` function

Reply:
I will try to test how my implementation behaves with larger bodies. Nice attempt though on re-implementing read. I would not have suspected that based on my assumptions of how reading works.

Reply:
> My assumption is that it updates the actual slice stored in SdkBody at that memory location.

This was my initial assumption as well, but more I looked into it the more I became unsure of it

Reply:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=c49aa3b8297bc690aacce83bcd2daef3

You are correct. I tested it and it does not replace the type stored there in sdkbody.
Above is my minimal reproduction.